pennylane/hf/hartree_fock.py

```diff
@@ -0,0 +1,195 @@
+            Fock matrix, core matrix
+        """
+        basis_functions = mol.basis_set
+        charges = mol.nuclear_charges
+        r = mol.coordinates
+        n_electron = mol.n_electrons
+
+        if r.requires_grad:
+            repulsion_tensor = generate_repulsion_tensor(basis_functions)(*args[1:])
+            s = generate_overlap_matrix(basis_functions)(*args[1:])
+        else:
+            repulsion_tensor = generate_repulsion_tensor(basis_functions)(*args)
+            s = generate_overlap_matrix(basis_functions)(*args)
+        h_core = generate_core_matrix(basis_functions, charges, r)(*args)
+
+        w, v = anp.linalg.eigh(s)
+        x = v @ anp.diag(anp.array([1 / anp.sqrt(i) for i in w])) @ v.T
+
+        v_fock, w_fock = anp.linalg.eigh(x.T @ h_core @ x)
+        coeffs = x @ w_fock
+
+        p = molecular_density_matrix(n_electron, coeffs)
+
+        for _ in range(n_steps):
+
+            j = anp.einsum("pqrs,rs->pq", repulsion_tensor, p)
+            k = anp.einsum("psqr,rs->pq", repulsion_tensor, p)
+
+            fock_matrix = h_core + 2 * j - k
+
+            v_fock, w_fock = anp.linalg.eigh(x.T @ fock_matrix @ x)
+
+            coeffs = x @ w_fock
+
+            p_update = molecular_density_matrix(n_electron, coeffs)
+
+            if anp.linalg.norm(p_update - p) <= tol:
+                break
+
+            p = p_update
+
+        return v_fock, coeffs, fock_matrix, h_core, repulsion_tensor
+
+    return scf
+
+
+def nuclear_energy(charges, r):
+    r"""Return a function that computes the nuclear-repulsion energy.
+
+    The nuclear-repulsion energy is computed as
+
+    .. math::
+
+        \sum_{i>j}^n \frac{q_i q_j}{r_{ij}},
+
+    where :math:`q`, :math:`r` and :math:`n` denote the nuclear charges, nuclear positions and
+    the number of nuclei, respectively.
+
+    Args:
+        charges (list[int]): nuclear charges
+        r (array[float]): nuclear positions
+
+    Returns:
+        function: function that computes the nuclear-repulsion energy
+
+    **Example**
+
+    >>> symbols  = ['H', 'F']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 2.0]], requires_grad = True)
+    >>> mol = Molecule(symbols, geometry)
+    >>> args = [mol.coordinates]
+    >>> e = nuclear_energy(mol.nuclear_charges, mol.coordinates)(*args)
+    >>> print(e)
+    4.5
+    """
+
+    def nuclear(*args):
+        r"""Compute the nuclear-repulsion energy.
+
+        Args:
+            args (array[array[float]]): initial values of the differentiable parameters
+
+        Returns:
+            array[float]: nuclear-repulsion energy
+        """
+        if r.requires_grad:
+            coor = args[0]
+        else:
+            coor = r
+        e = 0
+        for i, r1 in enumerate(coor):
+            for j, r2 in enumerate(coor):
+                if i > j:
+                    e = e + (charges[i] * charges[j] / anp.sqrt(((r1 - r2) ** 2).sum()))
```

Comment:
A comment here. Since all these objects are symmetric one could save some time just by adding the lower/upper triangular terms. However, I notice that both loops are running over all the coordinates, that's why you need the if's (I have heard they add significant overhead). Why not just to run over the indices of the lower triangular? am I missing something here?

Furthermore, why is the denominator squared? shouldn't this be:
 ```suggestion
                    e = e + (charges[i] * charges[j] / anp.sqrt(((r1 - r2)).sum()))
```
?

Reply:
Not sure about the square in the denominator, but indeed we should try to exploit symmetry if possible. Great suggestion Alain!

Reply:
`anp.sqrt(((r1 - r2)).sum()) ` gives the distance between the two position vectors, replaced it with `anp.linalg.norm(r1 - r2)` for more clarity.